Kitodo/src/main/java/org/kitodo/production/services/data/ProcessService.java

```diff
@@ -470,6 +472,7 @@ public List<Process> loadData(int offset, int limit, String sortField, SortOrder
                 .collect(Collectors.toList());
         query.restrictToProjects(projectIDs);
         query.defineSorting(SORT_FIELD_MAPPING.get(sortField), sortOrder);
+        query.performIndexSearches();
```

Comment:
The strategy here is to retrieve the process IDs for which a specific condition e.g. `stepopen:Archivierung` is met from the Search Index. Those IDs are used to extend the sql query with a IN-statement containing all the IDs retrieved from the search index. This can lead to REALLY large queries when you have 10.000 processes in your database and 9000 of them are in the state "Strukturierung". If you search for that stage the SQL IN clause will contain 9000 process IDs. With more processes in Kitodo the queries can become really large, maybe exhausting MySQL here.

Reply:
> Just an observation: The strategy here is to retrieve the process IDs for which a specific condition e.g. stepopen:Archivierung is met from the Search Index. Those IDs are used to extend the sql query with a IN-statement containing all the IDs retrieved from the search index

This was necessary in the past or with ES / OS as ES / OS can only retrieve JSON based data which are no objects and are not so good accessible. For this reason the ES / OS result was given to Hibernate to retrieve the objects from the database which could contain more information and accessing them was a lot easier. As the used ES / OS API could never result more than 10.000 hits there was a limit to the Hibernate query itself.

This is not needed anymore and should be removed. As after the query is send to Hibernate / Hibernate-Search the results are objects which are easy accessible and hold all information which are needed.

Reply:
We have a situation, where some information (metadata, but also filter settings like "stepopen") can only be queried using the index. So we need the index to answer the query we have to query the index somewhere in the process. In my review i will try to concentrate on functionality so that we can focus on performance later, but the general remark i have is this.

We have seen in different places (e.g. https://github.com/kitodo/kitodo-production/issues/4378) that raw SQL / a dedicated index queries often outperforms Hibernate standard queries, which do somethines hundred of subsequent queries to fetch data which could be fetched in one go. (If Hibernate supports the optimized SQL query)
 
The process list search is one of the most used features and it should be optimized as much as possible (withing the constraints of Hibernate) in my oppinion. When scrolling through the process list i still see A LOT of SQL queries. And queries like here

https://github.com/kitodo/kitodo-production/blob/b6fb9bac0836e9677d5d22b3693c229e79eac429/Kitodo/src/main/webapp/WEB-INF/templates/includes/processes/processActionsColumn.xhtml#L35

are executed for every single process leading to the classical n+1 query.

I am wondering wether we can not construct something like a ProcessListDTO Object (or ProcessListObject, this is not about the DTO concept), which contains only the information needed to construct the list. (and the statistics). The objects in the list already have the necessary properties like `hasChildren` precomputed, so there is no reason to do additional database queries to calculate them hundred of times. 

And since we are already using the index, why not **store** the necessary properties in the index. We do not have to store all projects associated with a process in the index, but we should store the information which is needed for display like `hasChildren`. If we have this, we ideally only need one query insted of hundreds to get all the data for the list and have good performance. This will also make issues like https://github.com/kitodo/kitodo-production/issues/6348 less annoying, although we can do something on usability there, too. 

The necessary features of Hibernate Search have already been referenced in other places

https://docs.jboss.org/hibernate/search/7.1/reference/en-US/html_single/#projections
https://docs.jboss.org/hibernate/search/7.1/reference/en-US/html_single/index.html#search-mapping-associated
https://docs.jboss.org/hibernate/search/7.1/reference/en-US/html_single/index.html#mapping-reindexing-derivedfrom

I think it is fantastic that we now reduce indexing time by only index the processes and using the Hibernate abstractions simplifies a lot of logic. But when necessary for performance it would be great if we could use dedicated custom logic to achieve the perfomance in places where it is mostly needed.